test/metrics_test.go

```diff
@@ -29,112 +30,114 @@ example.com:0 {
-	if err != nil {
-		t.Fatal(err)
-	}
+	metricName := "coredns_dns_response_rcode_count_total"
 
-	corefile := `org:0 {
-		auto {
-			directory ` + tmpdir + ` db\.(.*) {1}
-			reload 1s
-		}
-		prometheus localhost:0
-	}
+	corefile := `example.org:0 {
+	forward . 8.8.8.8:53
+	prometheus localhost:0
+}
 `
-
-	i, err := CoreDNSServer(corefile)
+	srv, err := CoreDNSServer(corefile)
+	defer srv.Stop()
 	if err != nil {
 		t.Fatalf("Could not get CoreDNS serving instance: %s", err)
 	}
 
-	udp, _ := CoreDNSServerPorts(i, 0)
+	udp, _ := CoreDNSServerPorts(srv, 0)
 	if udp == "" {
 		t.Fatalf("Could not get UDP listening port")
 	}
-	defer i.Stop()
-
-	// Write db.example.org to get example.org.
-	if err = ioutil.WriteFile(filepath.Join(tmpdir, "db.example.org"), []byte(zoneContent), 0644); err != nil {
-		t.Fatal(err)
+        if err != nil {
+		t.Fatalf("Could not get CoreDNS serving instance: %s", err)
 	}
-	// TODO(miek): make the auto sleep even less.
-	time.Sleep(1100 * time.Millisecond) // wait for it to be picked up
 
 	m := new(dns.Msg)
-	m.SetQuestion("www.example.org.", dns.TypeA)
+	m.SetQuestion("google.com.", dns.TypeA)
 
-	if _, err := dns.Exchange(m, udp); err != nil {
+	if _, err = dns.Exchange(m, udp); err != nil {
 		t.Fatalf("Could not send message: %s", err)
 	}
 
-	metricName := "coredns_dns_request_count_total" //{zone, proto, family}
-
-	data := test.Scrape("http://" + metrics.ListenAddr + "/metrics")
-	// Get the value for the metrics where the one of the labels values matches "example.org."
-	got, _ := test.MetricValueLabel(metricName, "example.org.", data)
-
-	if got != "1" {
-		t.Errorf("Expected value %s for %s, but got %s", "1", metricName, got)
+	// coredns cache size
+	cachSizeSamples := map[model.LabelName]model.LabelValue{
+		model.MetricNameLabel: model.LabelValue(metricName),
+		"server":              "dns://:0",
+		"rcode":               "REFUSED",
+		"zone":                "dropped",
+		"Value":               "1",
 	}
-
-	// Remove db.example.org again. And see if the metric stops increasing.
-	os.Remove(filepath.Join(tmpdir, "db.example.org"))
-	time.Sleep(1100 * time.Millisecond) // wait for it to be picked up
-	if _, err := dns.Exchange(m, udp); err != nil {
-		t.Fatalf("Could not send message: %s", err)
+	expect := test.DnsMetrics(
+		map[string]model.Samples{
+			metricName: model.Samples{test.NewCacheSample(model.LabelValue(metricName), cachSizeSamples)},
+		},
+	)
+	actual, err := test.GetDnsMetrics(map[string]interface{}{
+		metricName: nil,
+	}, metrics.ListenAddr)
+	if err != nil {
+		t.Errorf("Could not get dns metrics: %v", err)
 	}
-
-	data = test.Scrape("http://" + metrics.ListenAddr + "/metrics")
-	got, _ = test.MetricValueLabel(metricName, "example.org.", data)
-
-	if got != "1" {
-		t.Errorf("Expected value %s for %s, but got %s", "1", metricName, got)
+	t.Logf("current addr： %s\n", metrics.ListenAddr)
+	// compare to expected
+	if diff := cmp.Diff(expect, actual); diff != "" {
```

Comment:
cmd.Diff can be handy, but the old test clearly spelled out what was wrong, now this becomes more complex. I like the old way better.